### DISCUSSION TITLE
How to use maxgraph directly in a web page

### DISCUSSION THREAD

**Discussion by bochatay** [maxGraph/maxGraph#462] [Q&A] [answered]
Hello, I'd like to test maxgraph and I haven't succeeded.
In a directory shared with Apache (LAMP), I ran npm install @maxgraph/core
I added an index.html containing a div (id graph-container) and the script
I adapted the script to be javascript, as suggested in the readme
I added type=“module” to the script
`<div id="graph-container">`
`<script type="module">`
`import {type CellStyle, Graph, InternalEvent} from '@maxgraph/core';`
`const container = document.getElementById('graph-container');`
`...`
When I access to the page, the error is "Le spécificateur <@maxgraph/core> était un spécificateur simple, mais n'a pas été remappé. ..."
I think I need to use a node / npm feature I don't know about
Thanks in advance if you have any more details on how to make a "hello-world" with maxgraph

**Comment by reuse-ay** [comment]
How about deleting `type CellStyle`?

**Comment by tbouffard** [comment, marked as answer]
Hi @bochatay ,
As mentioned by @reuse-ay, you should remove the "type" directive, but the error is not coming from that.


### Error reproduction

I have reproduced your error.

Firefox 126.0.1 (French language)
```
Uncaught TypeError: Le spécificateur « @maxgraph/core » était un spécificateur simple, mais n’a pas été remappé.
Les spécificateurs relatifs de modules doivent commencer par « ./ », « ../ » ou « / ».
```

Chrome 126.0.6478.55
`Uncaught TypeError: Failed to resolve module specifier "@maxgraph/core". Relative references must start with either "/", "./", or "../".`

### Quick explanations

Imports using  "@maxgraph/core" are intended to be interpreted by bundlers, TypeScript, .... but don't work in the browser, which requires relative references (see the Chrome error) or absolute references.

The maxGraph npm package is intended to be used with a package manager and bundlers which will bundle maxGraph with the rest of your application (as explicitly states in the README).
The package doesn't provide a bundle for direct use in a browser (this has been removed in version 0.5.0, see #266) and there is no plan to change this AFAIK (but based on feedback, I guess that this could change).
So, for the time being, any solution that allows maxGraph to be used directly in a web page is a hack. Is there any reason why you can't integrate maxGraph into an application?


### Possible workaround/hack

That being said, I have tried to directly import the package entry point directly in the page. It is a ES module so it could be possible to load it.
This entry point (dist/index.js in the npm package) imports all maxGraph modules (classes, functions, constants, ...), so the browser should be theoretically able to load the whole maxGraph code.
Be aware that such a solution is very ineffective, as the browser is going to load a lot of files, and the files are not minified. So don't use it in production.

In the page, I changed the import directive like this:
```javascript
// use jsdelivr
import {Graph, InternalEvent} from 'https://cdn.jsdelivr.net/npm/@maxgraph/core@0.11.0/dist/index.js';
// or locally if the page is in the same folder as the 'node_modules' folder
// import {Graph, InternalEvent} from './node_modules/@maxgraph/core/dist/index.js';
```

The dist/index.js file imports elements without extension, so the browser tries to load https://cdn.jsdelivr.net/npm/@maxgraph/core@0.11.0/dist/view/mixins/ConnectionsMixin instead. And so, it fails to load the resources.

I finally found a workaround by locally patching the js files to ensure that all imports use the js extension to match the path of the file.
I have attached the solution in this zip (the content must be served by a web server, as the script type is "module" in the HTML page): 
[maxgraph_discussion_462_shared_archives_01.zip](https://github.com/user-attachments/files/15817357/maxgraph_discussion_462_shared_archives_01.zip)

The JavaScript code is the same as in the README of the maxGraph repo (after adaptation from TypeScript).

Here is what I then see in the browser 👇🏿 

![discussion_462_workaround_maxGraph_direcly_in_a_webpage](https://github.com/maxGraph/maxGraph/assets/27200110/b3ed676a-5cf2-4c6c-9a2a-1ab656a7b9b4)



**_Mini how-to patch the npm package_**
- Get the url the npm package (omit the version to get the latest): `$ npm view @maxgraph/core@0.11.0 dist.tarball`
--> https://registry.npmjs.org/@maxgraph/core/-/core-0.11.0.tgz
- Use the url to get the tgz
- Patch the file. I used https://github.com/beenotung/fix-esm-import-path: `npx fix-esm-import-path ./node_modules/@maxgraph/core/dist`
- Some files must be manually patched: they reference a folder instead of the index.js file that they contain. For example, in index.js replace `export * from './editor';` by `export * from './editor/index.js';`.
- Other imports to manually patch
  - index.js: from './serialization/codecs' to './serialization/codecs/index.js'
  - serialization/register.js: from './codecs' to './codecs/index.js'


### About the missing .js extension

This may be fixed in the future as it impacts some webpack configuration. For example, see https://github.com/maxGraph/maxGraph/blob/v0.11.0/packages/js-example/webpack.config.js#L21-L30.

**Comment by bochatay** [comment]
Hello,
Thank you very much for the explanations and your work to integrate maxGraph directly into a web page. The reason why I didn't use a bundler to create an application is that I didn't understand this prerequisite when I read the readme file. I've never created a node application. My project is a POC that I haven't started yet, so this is a good opportunity to learn.

**Reply by tbouffard** [reply]
Hi @bochatay 
I have created #490 to improve the getting started paragraph in the README
